File: src/csharp/4_BehavioralPatterns/5_Mediator/ChatRoom.cs
using System.Collections.Generic;
using System.Linq;
using static System.Console;

namespace DotNetDesignPatternDemos.Behavioral.Mediator.ChatRoom
{
  public class Person
  {
    public string Name;
    public ChatRoom Room;
    private List<string> chatLog = new List<string>();

    public Person(string name)
    {
      Name = name;
    }

    public void Receive(string sender, string message)
    {
      string s = $"{sender}: '{message}'";
      WriteLine($"[{Name}'s chat session] {s}");
      chatLog.Add(s);
    }

    public void Say(string message)
    {
      Room.Broadcast(Name, message);
    }

    public void PrivateMessage(string who, string message)
    {
      Room.Message(Name, who, message);
    }
  }

  public class ChatRoom
  {
    private List<Person> people = new List<Person>();

    public void Broadcast(string source, string message)
    {
      foreach (var p in people)
        if (p.Name != source)
          p.Receive(source, message);
    }

    public void Join(Person p)
    {
      string joinMsg = $"{p.Name} joins the chat";
      Broadcast("room", joinMsg);

      p.Room = this;
      people.Add(p);
    }

    public void Message(string source, string destination, string message)
    {
      people.FirstOrDefault(p => p.Name == destination)?.Receive(source, message);
    }
  }

  public class Demo
  {
    static void Main(string[] args)
    {
      var room = new ChatRoom();

      var john = new Person("John");
      var jane = new Person("Jane");

      room.Join(john);
      room.Join(jane);

      john.Say("hi room");
      jane.Say("oh, hey john");

      var simon = new Person("Simon");
      room.Join(simon);
      simon.Say("hi everyone!");

      jane.PrivateMessage("Simon", "glad you could join us!");
    }
  }
}

File: src/js/4_BehavioralPatterns/5_Mediator/chatroom.js
class Person
{
  constructor(name) {
    this.name = name;
    this.chatLog = [];
  }

  receive(sender, message)
  {
    let s = `${sender}: '${message}'`;
    console.log(`[${this.name}'s chat session] ${s}`);
    this.chatLog.push(s);
  }

  say(message) {
    this.room.broadcast(this.name, message);
  }

  pm(who, message)
  {
    this.room.message(this.name, who, message);
  }
}

class ChatRoom
{
  constructor()
  {
    this.people = [];
  }

  broadcast(source, message)
  {
    for (let p of this.people)
      if (p.name !== source)
        p.receive(source, message);
  }

  join(p)
  {
    let joinMsg = `${p.name} joins the chat`;
    this.broadcast('room', joinMsg);
    p.room = this;
    this.people.push(p);
  }

  message(source, destination, message)
  {
    for (let p of this.people)
      if (p.name === destination)
        p.receive(source, message);
  }
}

let room = new ChatRoom();

let john = new Person('John');
let jane = new Person('Jane');

room.join(john);
room.join(jane);

john.say('hi room');
jane.say('oh, hey john');

let simon = new Person('Simon');
room.join(simon);
simon.say('hi everyone!');

jane.pm('Simon', 'glad you could join us!');